File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        def isValidRec(root):
            if root == None :
                return [True, float("-inf"), float("inf")]
            
            leftRoot = root.left
            leftInfo = isValidRec(leftRoot)
            rightRoot = root.right
            rightInfo = isValidRec(rightRoot)

            maxLeft = leftInfo[1] 
            minRight = rightInfo[2]
            isGood = leftInfo[0] and rightInfo[0] and maxLeft < root.val < minRight

            newMaxLeft = max(root.val, rightInfo[1])
            newMinRight = min(root.val, leftInfo[2])

            return [isGood, newMaxLeft, newMinRight]
        val = isValidRec(root)
        return val[0]

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        int minVal = min(p->val, q->val);
    int maxVal = max(p->val, q->val);
    int currentVal = root->val;
    if (minVal <= currentVal && currentVal <= maxVal){
        return root; 
    }
    if (root->right != nullptr && maxVal > root->right->val){
        return lowestCommonAncestor(root->right, p, q);
    }
    return lowestCommonAncestor(root->left, p, q);
    }
};
